File: ts/src/scan.ts
/**
 * scan.ts — connect, pull tools/list, build a report card (token tax + safety +
 * progressive-disclosure estimate + cross-run rug-pull pinning).
 */

import { createHash } from "node:crypto";
import { readFileSync, writeFileSync, existsSync } from "node:fs";
import { McpClient, Tool } from "./client.js";
import { measure, countTokens, tokenizerName } from "./tokens.js";
import { highFindings, Finding } from "./forensics.js";

export interface ScanFinding { tool: string; severity: string; kind: string; detail: string; }
export interface ScanReport {
  server: string; transport: string; tokenizer: string; window: number;
  nTools: number; fullTokens: number; minimalTokens: number; ratio: number; pctContext: number;
  perTool: { name: string; tokens: number }[];
  findings: ScanFinding[];
  progressive: null | { indexTokens: number; baselineTokens: number; taskTokens: number; reductionPct: number };
  grades: { efficiency: string; safety: string };
  highFindings: number;
  rugPulls: number;
}

function fullSer(tools: Tool[]): string { return JSON.stringify(tools); }
function minimalSer(tools: Tool[]): string {
  return tools.map((t) => {
    const desc = (t.description ?? "").split(". ")[0].trim();
    const props = (t.inputSchema?.properties ?? {}) as Record<string, any>;
    const params = Object.entries(props).map(([k, v]) => `${k}:${v?.type ?? "any"}`).join(",");
    return `${t.name ?? "?"}(${params}) - ${desc}`;
  }).join("\n");
}

function digest(t: Tool): string {
  const canon = JSON.stringify({ name: t.name ?? null, description: t.description ?? null, inputSchema: t.inputSchema ?? null });
  return createHash("sha256").update(canon).digest("hex");
}

function grade(pct: number, high: number): { efficiency: string; safety: string } {
  const eff = pct < 5 ? "A" : pct < 12 ? "B" : pct < 20 ? "C" : "D";
  const safety = high === 0 ? "A" : high <= 2 ? "C" : "F";
  return { efficiency: eff, safety };
}

/** minimal progressive-disclosure model: index (2 meta-tools) + one loaded tool. */
function indexTokens(): number {
  const meta = [
    { name: "tool_search", description: "Search available tools by intent; returns ranked names and one-line summaries.", inputSchema: { type: "object", properties: { query: { type: "string", description: "What you want to do." } }, required: ["query"] } },
    { name: "tool_load", description: "Load the full schema for one tool by name.", inputSchema: { type: "object", properties: { name: { type: "string", description: "Exact tool name." } }, required: ["name"] } },
  ];
  return countTokens(JSON.stringify(meta));
}

export async function scan(client: McpClient, opts: { window?: number; pinPath?: string } = {}): Promise<ScanReport> {
  const window = opts.window ?? 200_000;
  const info = await client.initialize();
  const tools = await client.listTools();
  const server = info?.serverInfo?.name ?? "<unknown>";

  const full = tools.length ? measure(fullSer(tools)).tokens : 0;
  const minimal = tools.length ? measure(minimalSer(tools)).tokens : 0;
  const perTool = tools
    .map((t) => ({ name: t.name ?? "?", tokens: measure(JSON.stringify(t)).tokens }))
    .sort((a, b) => b.tokens - a.tokens);

  // safety: forensics on description + schema descriptions + name; plus rug-pull pinning
  const findings: ScanFinding[] = [];
  let pins: Record<string, string> = {};
  if (opts.pinPath && existsSync(opts.pinPath)) {
    try { const p = JSON.parse(readFileSync(opts.pinPath, "utf8")); if (p && typeof p === "object" && !Array.isArray(p)) pins = p; } catch { /* */ }
  }
  let rugPulls = 0;
  for (const t of tools) {
    const name = t.name ?? "<unknown>";
    const blobs: string[] = [t.description ?? "", t.name ?? ""];
    for (const v of Object.values((t.inputSchema?.properties ?? {}) as Record<string, any>)) {
      if (v && typeof v.description === "string") blobs.push(v.description);
    }
    if (t.annotations) blobs.push(JSON.stringify(t.annotations));
    for (const blob of blobs.map((b) => b.slice(0, 20000))) {
      for (const f of highFindings(blob)) findings.push({ tool: name, severity: "high", kind: kindOf(f), detail: `${f.technique}: ${f.evidence}` });
    }
    const d = digest(t);
    if (pins[name] && pins[name] !== d) { findings.push({ tool: name, severity: "high", kind: "rug-pull", detail: `definition changed since first seen (${pins[name].slice(0, 12)}… → ${d.slice(0, 12)}…)` }); rugPulls++; }
    pins[name] = d;
  }
  if (opts.pinPath) { try { writeFileSync(opts.pinPath, JSON.stringify(pins)); } catch { /* */ } }

  // dedupe findings by tool+detail
  const seen = new Set<string>();
  const deduped = findings.filter((f) => { const k = f.tool + "|" + f.detail; if (seen.has(k)) return false; seen.add(k); return true; });
  const high = deduped.filter((f) => f.severity === "high").length;

  let progressive: ScanReport["progressive"] = null;
  if (tools.length) {
    const baseline = full;
    const heaviestTask = perTool[0]?.tokens ?? 0;
    const idx = indexTokens();
    const searchResult = countTokens(JSON.stringify(perTool.slice(0, 3).map((t) => ({ name: t.name, summary: t.name }))));
    const taskTokens = idx + searchResult + heaviestTask;
    progressive = { indexTokens: idx, baselineTokens: baseline, taskTokens, reductionPct: baseline ? Math.round((1 - taskTokens / baseline) * 1000) / 10 : 0 };
  }

  return {
    server, transport: client.transport, tokenizer: tokenizerName(), window,
    nTools: tools.length, fullTokens: full, minimalTokens: minimal,
    ratio: minimal ? Math.round((full / minimal) * 100) / 100 : 0,
    pctContext: window ? Math.round((full / window) * 1000) / 10 : 0,
    perTool, findings: deduped, progressive, grades: grade(window ? (full / window) * 100 : 0, high),
    highFindings: high, rugPulls,
  };
}

function kindOf(f: Finding): string {
  const t = f.technique.toLowerCase();
  if (t === "injection") return "injection";
  if (t === "exfil") return "exfil";
  if (t.includes("homoglyph")) return "homoglyph";
  return "concealed-unicode";
}

File: mcpturn/scan.py
"""
scan.py — the scan engine: connect, pull tools/list, and build a report card.

Reuses the project's measurement + defense libraries so the CLI reports exactly
what the benchmark proves:
  - token tax (full vs. minimal-schema floor, per-tool, worst offenders)
  - metadata-safety findings (poisoning / concealment / rug-pull) via the
    i18n-honest mediator + unicode forensics
  - progressive-disclosure potential at the same boundary
  - optional cross-run digest pinning for real rug-pull detection over time
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from typing import Any

from defenses.mediator import CatalogPinner, mediate_catalog
from defenses.progressive import ProgressiveCatalog
from harness.mcpkit import count_tokens, measure, tokenizer_name


def _full(tools: list[dict]) -> str:
    return json.dumps(tools, separators=(",", ":"), ensure_ascii=False)


def _minimal(tools: list[dict]) -> str:
    lines = []
    for t in tools:
        desc = (t.get("description") or "").split(". ")[0].strip()
        props = (t.get("inputSchema") or {}).get("properties") or {}
        params = ",".join(f"{k}:{v.get('type','any')}" for k, v in props.items() if isinstance(v, dict))
        lines.append(f"{t.get('name','?')}({params}) - {desc}")
    return "\n".join(lines)


@dataclass
class ScanReport:
    server: str = ""
    transport: str = ""
    tokenizer: str = ""
    window: int = 200_000
    n_tools: int = 0
    full_tokens: int = 0
    minimal_tokens: int = 0
    ratio: float = 0.0
    pct_context: float = 0.0
    per_tool: list[dict] = field(default_factory=list)      # [{name,tokens}]
    findings: list[dict] = field(default_factory=list)      # [{tool,severity,kind,detail}]
    progressive: dict = field(default_factory=dict)
    grades: dict = field(default_factory=dict)

    @property
    def high(self) -> int:
        return sum(1 for f in self.findings if f["severity"] == "high")

    @property
    def rug_pulls(self) -> int:
        return sum(1 for f in self.findings if f["kind"] == "rug-pull")

    def to_dict(self) -> dict:
        d = {k: getattr(self, k) for k in (
            "server", "transport", "tokenizer", "window", "n_tools",
            "full_tokens", "minimal_tokens", "ratio", "pct_context",
            "per_tool", "findings", "progressive", "grades")}
        d["high_findings"] = self.high
        return d


def _grade(pct: float, high: int) -> dict:
    if pct < 5:
        eff = "A"
    elif pct < 12:
        eff = "B"
    elif pct < 20:
        eff = "C"
    else:
        eff = "D"
    sec = "A" if high == 0 else ("C" if high <= 2 else "F")
    return {"efficiency": eff, "safety": sec}


def scan(client, window: int = 200_000, pin_path: str | None = None,
         example_task: str = "create a new item") -> ScanReport:
    info = client.initialize()
    tools = client.list_tools()
    server = (info.get("serverInfo") or {}).get("name", "<unknown>") if isinstance(info, dict) else "<unknown>"

    rep = ScanReport(server=server, transport=getattr(client, "transport", "?"),
                     tokenizer=tokenizer_name(), window=window, n_tools=len(tools))

    # ---- token tax ----
    fm = measure(_full(tools)) if tools else {"tokens": 0}
    mm = measure(_minimal(tools)) if tools else {"tokens": 0}
    rep.full_tokens = fm["tokens"]
    rep.minimal_tokens = mm["tokens"]
    rep.ratio = round(fm["tokens"] / mm["tokens"], 2) if mm["tokens"] else 0.0
    rep.pct_context = round(100.0 * fm["tokens"] / window, 1) if window else 0.0
    rep.per_tool = sorted(
        ({"name": t.get("name", "?"), "tokens": measure(json.dumps(t, separators=(",", ":"), ensure_ascii=False))["tokens"]}
         for t in tools), key=lambda x: -x["tokens"])

    # ---- metadata safety (with optional cross-run pinning) ----
    pinner = CatalogPinner()
    if pin_path and os.path.exists(pin_path):
        try:
            with open(pin_path) as f:
                loaded = json.load(f)
            if isinstance(loaded, dict):  # ignore a hand-edited file of the wrong shape
                pinner._pins = loaded
        except Exception:
            pass
    result = mediate_catalog(tools, pinner=pinner, provenance=server)
    rep.findings = [{"tool": f.tool, "severity": f.severity, "kind": f.kind, "detail": f.detail}
                    for f in result.findings]
    if pin_path:
        try:
            with open(pin_path, "w") as f:
                json.dump(pinner._pins, f)
        except Exception:
            pass

    # ---- progressive-disclosure potential ----
    if tools:
        clean = [{k: v for k, v in t.items() if not k.startswith("_")} for t in result.cleaned_tools]
        pc = ProgressiveCatalog(clean)
        baseline = count_tokens(_full(clean))
        cost = pc.cost_for_task(example_task)
        rep.progressive = {
            "index_tokens": pc.index_tokens(),
            "baseline_tokens": baseline,
            "task_tokens": cost["tokens_paid"],
            "reduction_pct": round(100.0 * (1 - cost["tokens_paid"] / baseline), 1) if baseline else 0.0,
            "example_task": example_task,
            "loaded": cost["loaded"],
        }

    rep.grades = _grade(rep.pct_context, rep.high)
    return rep
